src/main/java/org/junitpioneer/jupiter/params/RangeSourceArgumentsProvider.java

```diff
@@ -41,20 +44,34 @@
  * @see DoubleRangeSource
  * @see FloatRangeSource
  */
-class RangeSourceArgumentsProvider implements ArgumentsProvider {
+class RangeSourceArgumentsProvider implements ArgumentsProvider, CartesianAnnotationConsumer<Annotation> {
+
+	private Annotation argumentsSource;
 
 	@Override
 	public Stream<? extends Arguments> provideArguments(ExtensionContext context) throws Exception {
 		// since it's a method annotation, the element will always be present
-		List<Annotation> argumentsSources = context
-				.getElement()
-				.map(method -> Arrays
-						.stream(method.getAnnotations())
-						.filter(annotations -> Arrays
-								.stream(annotations.annotationType().getAnnotationsByType(ArgumentsSource.class))
-								.anyMatch(annotation -> getClass().equals(annotation.value())))
-						.collect(Collectors.toList()))
-				.orElseThrow(IllegalStateException::new);
+		if (argumentsSource == null)
+			initArgumentsSource(context);
+		Class<? extends Annotation> argumentsSourceClass = argumentsSource.annotationType();
+		Class<? extends Range> rangeClass = argumentsSourceClass.getAnnotation(RangeClass.class).value();
+
+		Range<?> range = (Range<?>) rangeClass.getConstructors()[0].newInstance(argumentsSource);
+		range.validate();
+		return asStream(range).map(Arguments::of);
+	}
+
+	private void initArgumentsSource(ExtensionContext context) {
+		// since it's a method annotation, the element will always be present
+		AnnotatedElement element = context.getElement().orElseThrow(IllegalStateException::new);
+
+		verifyNoContainerAnnotationIsPresent(element);
+		List<Annotation> argumentsSources = Stream
+				.of(element.getAnnotations())
+				.filter(annotations -> Arrays
+						.stream(annotations.annotationType().getAnnotationsByType(ArgumentsSource.class))
+						.anyMatch(annotation -> getClass().equals(annotation.value())))
```

Comment:
Will this find meta-annotations? I doubt it would. Does this make sense?

```java
@IntRangeSource(from = 0, to = 10)
public @interface FromZeroToTen { }

@CartesianProductTest
@FromZeroToTen
void test(int i) {
}
```

If it does, we need to make sure that we find such annotations. Please use/extend `org.junitpioneer.jupiter.params.PioneerAnnotationUtils` if it's suitable - as painful as that may be.

Reply:
All range sources have `@Target(ElementType.METHOD)` on them. No meta-annotations can exist.

Reply:
True. We should probably discuss on an issue first whether this makes sense.